### ISSUE TITLE
Show payload being sent when using verbose with Invoke-RestMethod and Invoke-WebRequest

### ISSUE THREAD

**Issue by Nova1089** [PowerShell/PowerShell#23879]
### Summary of the new feature / enhancement

Please show the payload being sent when using verbose with Invoke-RestMethod and Invoke-WebRequest. Everything including the method, URI, headers, and body.

These cmdlets can be a nightmare to troubleshoot without inspecting the actual payload since they are doing some URL encoding, passing some defaults, etc. behind the scene.

**Comment by jhoneill** [comment]
This was changed in one of the versions of 7.  The argument was that "for security" (i.e. shut up and don't argue) the URL could not be shown - because some poorly implemented services put security info in the URL and setting verbose preference would cause some scripts to leak information.  I had to write a proxy function so I can see the URL being called when I'm trying to debug things.  The same argument will no doubt be thrown at any desire to see the payload, but you can use a proxy function to work around it.

**Comment by Nova1089** [comment]
> This was changed in one of the versions of 7. The argument was that "for security" (i.e. shut up and don't argue) the URL could not be shown - because some poorly implemented services put security info in the URL and setting verbose preference would cause some scripts to leak information.

How interesting.. not sure why you'd put sensitive info in the URI 😂 I suppose that HTTPS would encrypt that before it leaves your network, but many apps/browsers/clients will still capture that on the sending side. Besides, if I'm the one sending the data, should I not be privy to it? And if I'm writing an app where I want to keep it secret, I could just not expose it! Doesn't make much sense.

Could you share your proxy function? I'm not sure how to capture the payload being sent by Invoke-RestMethod and Invoke-WebRequest.

**Comment by SteveL-MSFT** [comment]
@PowerShell/wg-powershell-cmdlets discussed this, we agree on having the webcmdlets write out the request and response payloads as `debug` messages.  We defer to the @PowerShell/area-security WG to discuss next.

**Comment by SydneyhSmith** [comment]
As long as you need to opt in such as using `debug` @PowerShell/area-security WG does not have security concerns

**Comment by iSazonov** [comment]
I guess we want something like LoggingHandler? https://stackoverflow.com/questions/18924996/logging-request-response-messages-when-using-httpclient

**Comment by JustinGrote** [comment]
I think this should be 3-pronged:
- Debug output should emit two lines, one with the request URL, and one with the response status code.
- `-RequestVariable` parameeter that works like `-StatusCodeVariable`
- `-ResponseVariable` parameter that works like `-StatusCodeVariable`. This is a little tricky because the response is a stream, so it would have to be cloned before it is red.

I also agree with @iSazonov that a delegatinghandler (aka httpclient middleware) is the best way to handle this, it would basically be an augment to the existing one.

I'm willing to put together a POC PR if this is flagged `up-for-grabs`

**Comment by JustinGrote** [comment]
I read that it was approved as long as it is opt in, with debug being *one* such option.

The variable parameters are opt in as well, without them specified nothing happens. As far as security it's about info leakage to the console (api key in GET url for example) and those params dont enable that to happen.

**Comment by iSazonov** [comment]
@JustinGrote At first wg-powershell-cmdlets approved the `debug` messages and then Security WG only confirmed. If you want to add variables you should ask wg-powershell-cmdlets.

**Comment by JustinGrote** [comment]
I will start at that level then, the rest can be additive

**Comment by JustinGrote** [comment]
While investigating, I found it amusing that, if a link is a relative link, the original link gets exposed in the logs anyways. That was probably an oopsie :)

![Image](https://github.com/user-attachments/assets/2a9bda43-c65a-4b13-94da-45783b4be312)

**Comment by JustinGrote** [comment]
I prepared a [basic POC using builtin .NET methods](https://github.com/JustinGrote/_PR-PowerShell/tree/justingrote/feature/webRequestLogging)
![Image](https://github.com/user-attachments/assets/76053842-ef5d-451b-809a-1cc49fd7ddb3)


After some discussion on the #powershell-contributors Discord channel, some issues came up:
1. People currently using `-Debug` who aren't expecting anything today may accidentally leak secrets via GET URL query params in logs if they don't read the release notes.
1. Content logging for binary files is tricky, but most common byte types can be filtered so it doesn't fill the log with garbage for instance when -InFile is used.

I considered a REDACT for common headers similar to the Invoke-MgGraphRequest -Debug output but without knowing all possibilities this will just create a false sense of security.

I believe we can move forward here as-is, as long as it is heavily called out in the release notes, as I see the nightmare scenario being people who just have -Debug logging on, upgrade to PS7.6, and are suddenly exposing secrets via GET URL query parameters or request headers with no changes to their code.

@iSazonov @SydneyhSmith thoughts?

**Comment by iSazonov** [comment]
> While investigating, I found it amusing that, if a link is a relative link, the original link gets exposed in the logs anyways. That was probably an oopsie :)

It is for paging. In the case relative link has minor value.

> 1. People currently using `-Debug` who aren't expecting anything today may accidentally leak secrets via GET URL query params in logs if they don't read the release notes.

This problem has no general solution on any platform. This is why Windows logs are protected by access lists. The same must be observed for any transcript logs, including the PowerShell history log. Even a terminal itself could write a transcript with sensitive data.
The best thing to do is "security by default" what we provide with opt-in debug switch.

> Content logging for binary files is tricky, but most common byte types can be filtered so it doesn't fill the log with garbage for instance when -InFile is used.

If today we don't know best solution we can postpone this formatting to next iteration after getting feedback.
At least, I do not know that we have done anything to format the _response_ where this problem would appear more often.
So please start with simplest output.

I would start with the simplest output of _request_ in the default format (`list`). We can fast merge and get fast feedback.
Then (in follow PR) I would expect the headers and their values to be of the most importance during debugging and therefore should have a more convenient formatting for both the request (and the response).

**Comment by JustinGrote** [comment]
> I would start with the simplest output of _request_ in the default format (`list`). We can fast merge and get fast feedback. Then (in follow PR) I would expect the headers and their values to be of the most importance during debugging and therefore should have a more convenient formatting for both the request (and the response).

Sounds good. As for the binary outputs it's pretty easy to "allowlist" the known provider types and then for everything else just note that it is binary content. As the current implementation doesn't allow extensibility from other content types, this should be fairly stable.

**Comment by iSazonov** [comment]
> As for the binary outputs

For debugging I think it is more useful to see content encoding (MIME) type and content size than content itself.
The user knows what kind of content he is sending, but it is the erroneous type and too large a request size that may indicate the cause of the problem he is looking for.

**Comment by JustinGrote** [comment]
@iSazonov first thanks for the review, I'll work on the changes tomorrow.

As noted in the PR, there's some magic stuff that happens with the Body parameter and certain content types that might result in the actual request content looking different than what was supplied, which is why I thought it useful. Also, the body could be constructed programatically, and it provides an easy visual point to verify and inspect the call matches a REST API example. This is primarily for REST API POST calls, and not file uploads obviously.

**Comment by iSazonov** [comment]
@JustinGrote I agree that it can be useful. Maybe it's not that simple for output. That's why I'm talking about the priorities - size and type first. in
We could also think about the certificates used on the client and on the server.

**Comment by JustinGrote** [comment]
I like that idea! But agreed, start small and expand. PR updated accordingly
